### PR TITLE
xbps-install: add option to only download packages

### PR DESCRIPTION
**Pull request by Vaelatern**
I too have wanted, in the past, something like #8, for those times where I won't have a network later.

I want multiple people to OK this before I'll even consider merging. **This is untested.**

### CODE REVIEW

File: bin/xbps-install/main.c [hunk 1]
```diff
@@ -47,6 +47,7 @@ usage(bool fail)
 	    " -C --config <dir>        Path to confdir (xbps.d)\n"
 	    " -c --cachedir <dir>      Path to cachedir\n"
 	    " -d --debug               Debug mode shown to stderr\n"
+	    " -D --download-only       Download packages and check integrity, nothing else.\n"
```
Comment (chneukirchen) [inline bin/xbps-install/main.c]: Differently named long option below

File: include/xbps.h.in [hunk 2]
```diff
@@ -207,6 +207,13 @@
  */
 #define XBPS_FLAG_UNPACK_ONLY 		0x00001000
 
+/**
+ * @def XBPS_FLAG_DOWNLOAD_ONLY
+ * Do not configure packages in the transaction, just unpack them.
+ * Must be set through the xbps_handle::flags member.
+ */
+#define XBPS_FLAG_DOWNLOAD_ONLY 	0x00003000
```
Comment (chneukirchen) [inline include/xbps.h.in]: That flag value implies XBPS_FLAG_UNPACK_ONLY, why do you check both?

File: bin/xbps-install/main.c (hunk 1 above)
Comment (Vaelatern) [reply bin/xbps-install/main.c]: thanks, fixed.

File: include/xbps.h.in (hunk 2 above)
Comment (Vaelatern) [reply include/xbps.h.in]: Poorly thought out reasoning.

**Comment by CameronNemo** [comment]
It would be nice if the "Action" column was always set to "download" rather than "install", "reinstall", "downgrade", etc.

**Comment by Vaelatern** [comment]
Interesting, that would be good.

**Comment by CameronNemo** [comment]
I have been running this for a while, and nothing has broken. The functionality itself works well. The minor cosmetic issue I noted above should not be considered a blocker by any means.

**Comment by Vaelatern** [comment]
I'll make the cosmetic fix, and push to master

**Comment by CameronNemo** [comment]
Bump. Found a good use case for this.

**Comment by Vaelatern** [comment]
I think it does. Why?

**Comment by xtraeme** [comment]
Please repush

**Comment by CameronNemo** [comment]
I wanted to use this for my xmandump program but I think nilium is better equipped for that project, and xbps-query --cat may be better anyway.

**Comment by xtraeme** [comment]
Anybody interested on this one? last xbps version was released on July 2018, I'd like to have a new release before that date.

**Comment by kachsheev** [comment]
> Anybody interested on this one? last xbps version was released on July 2018, I'd like to have a new release before that date.

It's a interesting feature. I needed this one last week, and I was really surprised xbps doesn't have download-only mode.

**Comment by Vaelatern** [comment]
Re-pushed, just found this again.

**Comment by pullmoll** [comment]
The flag value should be `0x00002000`, though.

**Comment by Vaelatern** [comment]
I don't remember why but I wanted to trigger the UNPACK_ONLY code too.

File: include/xbps.h.in [hunk 3]
```diff
@@ -209,6 +209,13 @@
  */
 #define XBPS_FLAG_UNPACK_ONLY 		0x00001000
 
+/**
+ * @def XBPS_FLAG_DOWNLOAD_ONLY
+ * Do not configure packages in the transaction, just unpack them.
```
Comment (xtraeme) [inline include/xbps.h.in]: This description is misleading.

**Comment by pullmoll** [comment]
@Vaelatern  IMO you should in that case combine the flags `XBPS_FLAG_UNPACK_ONLY | XBPS_FLAG_DOWNLOAD_ONLY` when the `-D` option is given.

Or at the very least you should 
```
#define XBPS_FLAG_DOWNLOAD_ONLY 	(0x00002000 | XBPS_FLAG_UNPACK_ONLY)
```
I think everyone expects a flag to be a flag bit and not a combination of previously defined bits.

FWIW when I define flags like these I prefer the notation
```
#define FLAG1  (1 << 0)
#define FLAG2  (1 << 1)
…
#define FLAG16 (1 << 15)
```
because it's easier to read and slightly simpler to modify in case you want to insert a new definition before others.

File: include/xbps.h.in (hunk 3 above)
Comment (Vaelatern) [reply include/xbps.h.in]: Fixed.

**Comment by Vaelatern** [comment]
Fixed the flag, pullmoll.

Added flag documentation.

File: lib/transaction_commit.c [hunk 4]
```diff
@@ -265,6 +265,11 @@ xbps_transaction_commit(struct xbps_handle *xhp)
 		    "%s\n", strerror(rv));
 		goto out;
 	}
+
+	if (xhp->flags & XBPS_FLAG_DOWNLOAD_ONLY) {
+		goto out;
+	}
+
```
Comment (Vaelatern) [inline lib/transaction_commit.c]: I'm pretty sure I use the UNPACK_ONLY code in this function, and use this to shortcut and get out.

Comment (Vaelatern) [reply lib/transaction_commit.c]: Now that you mention it, @pullmoll, had we pushed with 0x00003000 this would have broken unpacking.

Comment (pullmoll) [reply lib/transaction_commit.c]: *Always be careful with the clamps. You may short circuit something you did not intend.*

**Pull request closed without merging**